Docs/monorepo (#248)
SenseNet/sensenet.github.io: _posts/2019-01-10-moving-client-packages-to-monorepo.md

﻿---
title: 'Moving sensenet client packages to monorepo'
author: B3zo0
image: '../img/posts/monorepo/colorful-stack.jpg'
tags: [monorepo, refactor, JavaScript, client packages]
---

_Juggling a multimodule project over multiple repos is like trying to teach a newborn baby how to ride a bike._ [Babel](https://github.com/babel/babel/blob/master/doc/design/monorepo.md#why-is-babel-a-monorepo)

---

## TL;DR

> *We moved all our javascript packages to [one reposity](https://github.com/SenseNet/sn-client).*

Reviewer: ```suggestion
*We moved all our javascript packages to [one repository](https://github.com/SenseNet/sn-client).*
```

## Some history about our client packages

> At first, we started with one package [sn-client-js](https://github.com/SenseNet/sn-client-js). As time goes by it grow bigger and we felt like we need to do something.

Reviewer: ```suggestion
At first, we started with one package [sn-client-js](https://github.com/SenseNet/sn-client-js). As time went by it grew bigger and we felt like we need to do something.
```

## Monorepo to the rescue

> In revision control systems, a monorepo (syllabic abbreviation of monolithic repository) is a software development strategy where code for many projects are stored in the same repository.

**In a monorepo the whole workflow can be simplified like this:**

1.  Create a branch from develop
2.  Create the function
3.  Build, test
4.  Merge back when done

So much easier and cleaner than before. No need to copy anything, anywhere. No need to release a dev alpha or a minor version. You can see your changes in all the packages once you modified something. It not only made our life simpler but it gave us a productivity boost as well. Let's see how the transition went.

![Patrick](/img/posts/monorepo/patrick-meme.jpg 'Get all the packages we extracted from one repository and put them together in a new one')

## Moving sensenet client packages to monorepo

First thing first, we need a new reopsitory. We could also use one of the repos and put everything there, but it felt cleaner to start a new one. I have created a todo list in this [pull request](https://github.com/SenseNet/sn-client/pull/1) to track the progress and to see what is left. 

> We moved all the packages to the packages folder with lerna. _[Lerna](https://lernajs.io/) is a tool that optimizes the workflow around managing multi-package repositories with git and npm._ This way we could see the commit history in the new repository as well. We had to make sure that every package is able to build itslef. We faced with some problems here. Not every package was configured correctly. Some was missing dependencies that were needed for it to be able to build.

Reviewer: ```suggestion
We moved all the packages to the packages folder with lerna. _[Lerna](https://lernajs.io/) is a tool that optimizes the workflow around managing multi-package repositories with git and npm._ This way we could see the commit history in the new repository as well. We had to make sure that every package is able to build itslef. We faced some problems here. Not every package was configured correctly. Some was missing dependencies that were needed for it to be able to build.
```

> After that we run the tests to make sure we didn't break anything. Because the repositories were not consistent in their configs, it failed. We had to temporarily suspend the linting to see the test results. With that said we realized that we should to make all repositories consistent. Fortunately there were no big errors.

Reviewer: ```suggestion
After that we run the tests to make sure we didn't break anything. Because the repositories were not consistent in their configs, it failed. We had to temporarily suspend the linting to see the test results. With that said we realized that we should make all repositories consistent. Fortunately there were no big errors.
```
